### PR TITLE
DEX-483 Testcontainers and IT API refactoring

### PR DESCRIPTION
**Pull request by DenisSerebryansky**
Testcontainers:

 * BaseContainer built upon GenericContainer introduced (DexContainer and WavesNodeContainer)
 * BaseContainerInfo introduced
 * REST API integrated into BaseContainer
 * ToxiProxy container used for connection issues simulation

IT API refactoring:

 * New MatcherSuiteBase is used in all IT tests:
     - all logic related to DEX and Node extracted into HasDex and HasWavesNode
     - integrated REST API usage: no more startAndAwait, waitReady, dex1Api and wavesNode1Api!
     - human-readable way to restart container with new suite config
     - all cache invalidations encapsulated into start/restart/connectToNetwork methods
     - spotify docker client isn't used anymore, all it's logic encapsulated into BaseContainer trait
     - coevals removed, using lazy vals for containers and run configs instead
 * Common logic for logging of the tests start extracted to InformativeTestStart
 * ItSuiteBase renamed to IntegrationSuiteBase (uses all features of new MatcherSuiteBase)
 * dex-it-tools module renamed to dex-it-common
 * dex-it-common now has dex module as dependency
 * DexApiOps, FailWith, ItMatchers, all DEX's and Node's responses moved to dex-it-common module
 * Warn logs from WavesBlockchainGrpcAsync/WavesBlockchainCachingClient removed
 * OrderHistoryTestSuite refactored:
     - Testcontainer of Postgres used
     - retry method removed
     - diffx is used
 * Spotify dependency, Docker class and its users removed
 * DoubleApiOps now uses model's Normalization
 * waitForOrderPlacement method for DexApi introduced

Miscellaneous:

 * Duplicated setProperty removed from Application
 * DEX extension port in application.conf fixed
 * Redundant port setting for DEX extension removed from dex-base.conf
 * Default RepeatRequestOptions (30 sec, waitReady uses 60 sec) are used in repeatUntil in FOps
 * Unused com.wavesplatform.dex.it.fp.Logger/DockerContainerLauncher/init-user-db.sh/com.wavesplatform.it.sync.smartcontracts package object removed

### CODE REVIEW

**Comment by DenisSerebryansky** [comment]
TODO:
 - Migrate the rest of the TetsSuites to the new MatcherSuiteBase and remove old one
 - Remove spotify dependency
 - Investigate possibility to avoid using of Testcontainers inner docker client
 - Suggest ways of IT API additional refactoring: more clear usage of api and containers, maybe get rid of caches

File: dex-it/src/test/scala/com/wavesplatform/it/sync/DEXClientFaultToleranceTestSuite.scala [hunk 1]
```diff
@@ -3,63 +3,75 @@ package com.wavesplatform.it.sync
...
-    dockerClient.disconnectFromNetwork(dex1Container)
+    proxy.setConnectionCut(true)
 
-    broadcastAndAwait(wavesNode1Api, alice2BobTransferTx)
-    usdBalancesShouldBe(wavesNode1Api, 0, defaultAssetQuantity)
+    broadcastAndAwait(wavesNodeApi, alice2BobTransferTx)
+    usdBalancesShouldBe(wavesNodeApi, 0, defaultAssetQuantity)
 
-    Thread.sleep(2000)
+    Thread.sleep(5000)
```
Comment (ngadiyak) [inline dex-it/src/test/scala/com/wavesplatform/it/sync/DEXClientFaultToleranceTestSuite.scala]: could we exclude this method of waiting from tests?

File: dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala [hunk 2]
```diff
@@ -1,95 +1,84 @@
...
+       |  data-source-class-name = "org.postgresql.ds.PGSimpleDataSource"
+       |}
+    """.stripMargin
 
-    postgresContainerLauncher.writeFile(
-      to = Paths.get(s"/docker-entrypoint-initdb.d/$createDatabaseFileName"),
-      content = createDatabaseFileContent
-    )
-  }
+  private val postgres: PostgreSQLContainer =
```
Comment (ngadiyak) [inline dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala]: maybe we should move this code to the base suite and add the trait with the DB cleanup? or move it to separate class and when we need the postgress in some test, write an annotation

File: dex-it/src/test/scala/com/wavesplatform/it/sync/DEXClientFaultToleranceTestSuite.scala (hunk 1 above)
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/DEXClientFaultToleranceTestSuite.scala]: We use this in order to simulate 5 seconds connection loss

File: dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala (hunk 2 above)
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala]: We use Postgres only in this test suite. We don't need ad-hoc trait with Postgres I guess

File: dex-it/src/test/scala/com/wavesplatform/it/sync/DEXClientFaultToleranceTestSuite.scala (hunk 1 above)
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/DEXClientFaultToleranceTestSuite.scala]: Done

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/DexApi.scala [hunk 3]
```diff
@@ -39,6 +39,7 @@ trait DexApi[F[_]] extends HasWaitReady[F] {
 
   def tryPlace(order: Order): F[Either[MatcherError, MatcherResponse]]
   def tryPlaceMarket(order: Order): F[Either[MatcherError, MatcherResponse]]
+  def retryPlace(order: Order): F[Either[MatcherError, MatcherResponse]]
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/DexApi.scala]: Done

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/HasDEX.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import com.wavesplatform.dex.it.docker.base.{BaseContainer, DEXContainer}
+import com.wavesplatform.dex.it.fp
+import com.wavesplatform.dex.it.fp.CanExtract
+import monix.eval.Coeval
+import mouse.any._
+
+import scala.concurrent.Future
+import scala.util.Try
+
+trait HasDEX { self: BaseContainersKit =>
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/HasDEX.scala]: Done :pepe:

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/HasDEX.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  protected val dexContainer: Coeval[DEXContainer] = Coeval.evalOnce { createDEX("dex-1") }
+
+  protected val cachedDEXApiAddress = CachedData(dexContainer().restApiAddress)
+
+  protected def dexApi: DexApi[Id]          = fp.sync { DexApi[Try](apiKey, cachedDEXApiAddress.get()) }
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/HasDEX.scala]: Done :pepe:

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/node/WavesNodeApiExtensions.scala [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
 trait WavesNodeApiExtensions {
+
   this: HasWavesNode =>
 
   protected def broadcastAndAwait(txs: Transaction*): Unit = broadcastAndAwait(wavesNode1Api, txs: _*)
 
   protected def broadcastAndAwait(wavesNodeApi: NodeApi[Id], txs: Transaction*): Unit = {
-    txs.map(wavesNodeApi.broadcast)
-    txs.foreach(tx => wavesNodeApi.waitForTransaction(tx))
+    txs map wavesNodeApi.broadcast
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/node/WavesNodeApiExtensions.scala]: Done

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/docker/base/BaseContainer.scala [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+    val localDestination = localLogsDir.resolve(s"container-$name.log")
+
+    log.info(s"$prefix Writing log to '${localDestination.toAbsolutePath}'")
+
+    val openOption = if (Files exists localDestination) StandardOpenOption.APPEND else StandardOpenOption.CREATE
+
+    Files.write(localDestination, container.getLogs.replace("\n\n", "\n") getBytes StandardCharsets.UTF_8, openOption)
+
+    this match {
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/docker/base/BaseContainer.scala]: Done

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/docker/base/info/DEXContainerInfo.scala [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+  val restApiPort: Int       = 6886 // application.conf waves.dex.rest-api.port
+  val exposedPorts: Seq[Int] = Seq(restApiPort)
+  val netAlias: String       = "d3x"
+
+  def getEnv(containerName: String): Map[String, String] = Map(
+    "WAVES_DEX_CONFIGPATH" -> s"$baseContainerPath/$containerName.conf",
+    "WAVES_DEX_OPTS" -> Seq(
+      "-Dlogback.stdout.enable=false",
+      "-Dlogback.file.enable=false",
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/docker/base/info/DEXContainerInfo.scala]: Done

File: dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala [hunk 9]
```diff
@@ -251,27 +205,27 @@ class OrderHistoryTestSuite extends MatcherSuiteBase {
...
-    dex1Api.cancel(alice, buyOrder)
+    dex1.api.cancel(alice, buyOrder)
 
-    retry(10, batchLingerMs) {
+    eventually {
       withClue("checking info for 2 small submitted orders\n") {
 
         Set(sellOrder1, sellOrder2).foreach { order =>
-          getOrderInfoById(order.id()).get shouldBe
+          getOrderInfoById(order.id()).get shouldBe // TODO DIFFX
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala]: diffx? :)

Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala]: It was not so easy, but it's done

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/node/NodeApi.scala [hunk 10]
```diff
@@ -120,7 +120,7 @@ object NodeApi {
           case NonFatal(e)                               => M.raiseError(e)
         }
 
-        repeatUntil(request, RepeatRequestOptions(1.second, 60))(_ == true).map(_ => ())
+        repeatUntil(request)(_ == true).map(_ => ())
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/node/NodeApi.scala]: Done

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/DexApiOps.scala [hunk 11]
```diff
@@ -1,11 +1,11 @@
-package com.wavesplatform.it.api
+package com.wavesplatform.dex.it.api.dex
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/DexApiOps.scala]: In sake of arrangement

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/DexApi.scala [hunk 12]
```diff
@@ -331,7 +333,7 @@ object DexApi {
         }
 
         // Sometimes container start during 20 seconds! https://github.com/docker/for-mac/issues/1183
-        repeatUntil(request, RepeatRequestOptions(1.second, 60))(_ == true).map(_ => ())
+        repeatUntil(request)(_ == true).map(_ => ())
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/DexApi.scala]: Done

**Pull request merged**